src/uoapi/course/parse.py

```diff
@@ -0,0 +1,111 @@
+    Subject,
+)
+
+from typing import (
+    cast,
+)
+
+
+def title_tag(tag: Tag | NavigableString) -> tuple[str, str, int]:
+    """
+    Extracts course code, title, and credits from a courseblocktitle tag
+    """
+    title = utils.replace_special_spaces(tag.text)
+
+    code = utils.extract_codes(title, False)[0]
+    title = utils.remove_codes(title)
+
+    credits = utils.extract_credits(title)
+    title = utils.remove_credits(title)
+
+    if (course_match := pt.code_groups.search(code)) is None:
+        raise ValueError(f"Could not parse course code {code}")
+
+    code = cast(str, course_match.groups()[1]).upper()
+
+    return code, title, credits
+
+
+def description_tag(tag: Tag | NavigableString | None) -> str:
+    """
+    Extracts the description of a course from a courseblockdesc tag
+    """
+    if tag is None:
+        return ""
+
+    description = utils.replace_special_spaces(tag.text)
+    return description
+
+
+def subject_tag(tag: Tag, url_prefix: str):
+    if not tag.has_attr("href"):
+        # TODO: Add log message here or crash
+        return None
+
+    match tag.string, tag["href"]:
+        case str(label), str(href):
+            path = utils.get_last_path_component(href)
+            subject = utils.clean_subject_label(label)
+            subject_code = path.upper()
+
+            return Subject(
+                subject=subject,
+                subject_code=subject_code,
+                link=url_prefix + path + "/",  # pyright: ignore
+            )
+        case s, h:
+            raise ValueError(f"Expected strings, got {type(s)} and {type(h)}")
+
+
+def extras_blocks(tags: list[Tag]) -> tuple[str, str]:
+    """
+    Extracts the prerequisites and components
+    from a list of courseblockextra tags
+
+    Args:
+        tags: The list of courseblockextra tags
+
+    Returns:
+        A tuple containing the prerequisites and component strings.
+        If the prerequisites or components are not found, an empty string
+        is returned for that value.
+    """
+    blocks: list[str] = []
+
+    for tag in tags:
+        blocks.append(
+            # No idea if the additional strips are necessary
+            # but they were in the original code so I'm keeping them
+            # for now
+            utils.replace_special_spaces(tag.text)
+            .strip(".")
+            .strip()
+            .strip(".")
+            .strip()
+        )
+
+    match blocks:
+        case [block] if utils.has_component(block):
+            return "", block
+        case [block] if utils.has_prerequisite(block):
+            return block, ""
+        case [pre, comp] | [comp, pre] if utils.is_prereq_and_component(pre, comp):
+            return pre, comp
+        case [block, _] | [_, block] if utils.is_prerequisite_string(block):
+            return block, ""
+        case [block, _] | [_, block] if utils.is_component_string(block):
+            return "", block
+        case _:
+            pass
```

Comment:
Is there a reason not to do the following?
```suggestion
    match blocks:
        case [pre, comp] | [comp, pre] if utils.is_prereq_and_component(pre, comp):
            return pre, comp
        case [block] |  [block, _] | [_, block] if utils.is_prerequisite_string(block):
            return block, ""
        case [block] | [block, _] | [_, block] if utils.is_component_string(block):
            return "", block
        case _:
            pass
```

I'm willing to believe the old code omitted the `not has_prerequisite(string)` check in `utils.is_component_string(block)` in the case of 1 block.
I'm just wondering if you've thought about whether this is the desired behaviour or not.
I'm not sure if that would affect the parsing of any courses.

Reply:
I wasn't sure if that would impact any of the course parsing, so I made sure to keep the same logic. I do quite prefer the legibility of your suggested one though, so if folks are fine with it I'd prefer to use that over the current longer match statement

Reply:
So, in light of the bug below, I think we should maybe assemble a bunch of test cases to prevent further issues. Specifically, grab a couple of subjects' pages (hopefully with "broad" coverage), run the parser as-is, verify their outputs, and make those "assert expected_output == parse(corresponding_input)" tests.
(However I believe this may be a fair amount of manual work.)

Reply:
Also wanted to echo @DudeRandom21 's comment: the match is orders of magnitude clearer than the former convoluted logic.

Reply:
> So, in light of the bug below, I think we should maybe assemble a bunch of test cases to prevent further issues. Specifically, grab a couple of subjects' pages (hopefully with "broad" coverage), run the parser as-is, verify their outputs, and make those "assert expected_output == parse(corresponding_input)" tests. (However I believe this may be a fair amount of manual work.)

Added #21 to track the effort to do so, if possible I'd like to tackle that as a separate PR once I've fixed the identified regression in this current PR